### DISCUSSION TITLE
Help with Frigate in an LXC with Proxmox

### DISCUSSION THREAD

**Discussion by andynbaker** [blakeblackshear/frigate#451] [Ask A Question] [answered]
Hey guys, I've been toying with proxmox and virtualization on another box (pfsense) and have been considering converting my ubuntu server running my home assistant stack (and frigate) over to a virtualized home assistant OS environment under proxmox. Before I take the plunge, can anyone help me out here?

If I understand it correctly, I would use proxmox to spin up a debian LXC on an amd-64 box. I would bind mount my coral usb dongle to the container, give it a nice chunk of RAM and a few cores, and then install docker and docker-compose in this LXC. From there, I'd spin up frigate in docker just like I'm doing now in bare metal ubuntu-server? Am I missing something? 

Edit:
I found this article helpful for figuring out the correct mappings for my usb coral: https://medium.com/@konpat/usb-passthrough-to-an-lxc-proxmox-15482674f11d

**Comment by TheSleepySlee** [comment, marked as answer]
This is what I & others are doing. Look at notes in the readme to have some config details.

**Comment by andynbaker** [comment]
Got it up and running this morning. Not too bad. The stinking coral mount point keeps changing every time I fiddle with it, though. Anyone know how to bind the device by id instead of by path?

**Reply by TheSleepySlee** [reply]
Don't move the coral and port numbers should not change. Using ID does not work good.

**Reply by andynbaker** [reply]
It was strange. Even just unplugging it and plugging it back into the same port bumped the numbers. I know what to look for now, so no biggie.

**Comment by thundercat49** [comment]
would u mind writing up a short guide on running in an LXC on promox or  and share your config?, do u use debian , or the turnkey version.. i tried a couple of times now and failed ( mental block around docker compose . setting up shares etc) 

am presently using the HA version in a VM on proxmox on an NUC, been really struggling to get the beta to work - got it recording clips, lots of error messages,  now coral is playing up, with it defaulting back to CPU< so i would like to try just running frigate in an LXC  on its own...see if its more stable

**Comment by andynbaker** [comment]
I can't put together anything super detailed, but the basic steps are as follows:

1. Install Proxmox hypervisor on bare metal.

2. Stand up a linux LXC (i'm using ubuntu 20.04 because that's what i'm used to, but debian or anything else probably works). Check out [this ](https://www.youtube.com/watch?v=1EYAGl96dZY)video on the process, it's pretty easy.

3. Plug in your coral and use the proxmox LXC console to enter the container and [install the coral libraries, etc.](https://coral.ai/docs/accelerator/get-started/?gclid=CjwKCAiArbv_BRA8EiwAYGs23JixmP93AmpQbvvoneLAhy6Z-YeWBX3O1O51MwJyVtJrsieHzaMc3RoCExUQAvD_BwE). Go all the way to the end and try to run the parrot test. It will fail.

4. We need to pass the usb device through to the container. Using [this](https://medium.com/@konpat/usb-passthrough-to-an-lxc-proxmox-15482674f11d) as a guide, find the device mapping. Use the proxmox console to enter the shell of the proxmox host (the node, not the container). Run the `lsusb` command to find your coral. It may or may not be listed as a google device. Mine showed up first as some generic thing, but then changed to a google device after I installed the coral libraries. You need the device path (from `lsusb`).

5. While you are in the host shell, edit the LXC config file with `nano /etc/pve/lxc/<container_id>.conf`. Add the lines at the bottom per the [frigate docs](/etc/pve/lxc/<container_id>.conf). Replace the device path and major and minor numbers with the values you found. Reboot the container and repeat step 4 to verify. I found the device path changed every time i unplugged the device and plugged it back in. It also seems to change if you add a new device in the future, so get used to this. Update the values in the container config if they changed, and reboot the guest.

![image](https://user-images.githubusercontent.com/53503316/103444706-aa0c3500-4c30-11eb-9796-11243d509980.png)

6. From here on out, you can treat the new container like any bare metal linux server on your network. [Install docker](https://linuxize.com/post/how-to-install-and-use-docker-on-ubuntu-20-04/) and [docker-compose](https://www.digitalocean.com/community/tutorials/how-to-install-and-use-docker-compose-on-ubuntu-20-04). Install portainer if that's your thing (I can't live without it), and then install frigate in docker in your new LXC server. If you are using portainer, it is super easy to watch the logs while the container starts up. You should be looking for the frigate container to find coral in the logs. If it cant find it, go back to step 4 and try again.

![image](https://user-images.githubusercontent.com/53503316/103444676-54378d00-4c30-11eb-9825-c48be13a3648.png)

7. Other tips and tricks: be sure to assign a static IP address using your method of choice. I'm running pfsense in another vm and added a DHCP record for my new frigate container. You can also assign it a static address when setting it up in proxmox. Pick your poison.

Here's my final config. I have 8 cameras running detection on the substreams (700x480) and restreaming the 1080p streams. 2G of RAM seems to be plenty:

```
arch: amd64
cores: 2
features: nesting=1
hostname: frigate
memory: 2048
net0: name=eth0,bridge=vmbr0,firewall=1,hwaddr=7E:E9:2D:57:4E:EB,ip=dhcp,ip6=dhcp,type=veth
onboot: 1
ostype: ubuntu
rootfs: local-lvm:vm-101-disk-0,size=12G
swap: 512
lxc.cgroup.devices.allow: c 189:* rwm
lxc.mount.entry: /dev/dri/renderD128 dev/dri/renderD128 none bind,optional,create=file
lxc.mount.entry: /dev/bus/usb/004/003 dev/bus/usb/004/003 none bind,optional,create=file
lxc.apparmor.profile: unconfined
lxc.cgroup.devices.allow: a
lxc.cap.drop:
```

**Reply by thundercat49** [reply]
many thanks, will give it a go this evening:)#

**Reply by andynbaker** [reply]
Quick update. I found that you can pass all minor numbers through by using an '*" instead of typing in the actual minor number. So the container config should look like this instead:

```
lxc.cgroup.devices.allow: c 189:* rwm
lxc.mount.entry: /dev/dri/renderD128 dev/dri/renderD128 none bind,optional,create=file
lxc.mount.entry: /dev/bus/usb/004/003 dev/bus/usb/004/003 none bind,optional,create=file
lxc.apparmor.profile: unconfined
lxc.cgroup.devices.allow: a
lxc.cap.drop: 
```
That "*" in the first line means you no longer need to find the minor number, just the path. One less thing to deal with.

**Reply by andynbaker** [reply]
Did you triple check your lxc config is correct? If so, check it again. My device path changed again this morning after a reboot. I also never successfully completed the parrot test. Go into the lxc shell and run lsusb and make sure the coral shows up.

**Reply by flaektrem** [reply]
> 
> 
> Did you triple check your lxc config is correct? If so, check it again. My device path changed again this morning after a reboot. I also never successfully completed the parrot test. Go into the lxc shell and run lsusb and make sure the coral shows up.


Sorry for messing up last reply, my bad.

My LXC lsusb:
````
root@Ubuntu:~# lsusb
Bus 002 Device 007: ID 0a5c:5800 Broadcom Corp. BCM5880 Secure Applications Processor
Bus 002 Device 006: ID 413c:8187 Dell Computer Corp. DW375 Bluetooth Module
Bus 002 Device 005: ID 1a86:7523 QinHeng Electronics HL-340 USB-Serial adapter
Bus 002 Device 004: ID 1cf1:0030 Dresden Elektronik 
Bus 002 Device 009: ID 1a6e:089a Global Unichip Corp. 
Bus 002 Device 002: ID 8087:0020 Intel Corp. Integrated Rate Matching Hub
Bus 002 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 001 Device 003: ID 05ca:1814 Ricoh Co., Ltd HD Webcam
Bus 001 Device 002: ID 8087:0020 Intel Corp. Integrated Rate Matching Hub
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
````

And my LXC config:
````
arch: amd64
cores: 2
features: nesting=1
hostname: Ubuntu
memory: 2000
net0: name=eth0,bridge=vmbr0,firewall=1,hwaddr=FE:3D:1E:45:23:A9,ip=dhcp,type=veth
ostype: ubuntu
rootfs: local-lvm:vm-104-disk-0,size=20G
swap: 512
unprivileged: 1
lxc.cgroup.devices.allow: c 189:* rwm
lxc.mount.entry: /dev/dri/renderD128 dev/dri/renderD128 none bind,optional,create=file
lxc.mount.entry: /dev/bus/usb/002/009 dev/bus/usb/002/009 none bind,optional,create=fi$
lxc.apparmor.profile: unconfined
lxc.cgroup.devices.allow: a
lxc.cap.drop:
````

**Reply by andynbaker** [reply]
And you are sure your major number is 189? Id reboot the host again for good measure and then start checking your frigate config. Also, mine changed from global unichip to Google in lsusb after passing it through and rebooting. Not sure if that matters.

**Reply by flaektrem** [reply]
Major number is 189 yes. But now while i was away for 5 minutes it changed to 002 and 010. It also changed from Global Unichip to Google...
This is soo strange!

So you also failed the parrot test?

**Reply by TheSleepySlee** [reply]
Never tried the parrot test.
Make sure your USB port has enough power. I see Google in the lsusb in the docker LXC.

Check usb errors here:
dmesg|grep usb

Also I have this set in LXC conf:
unprivileged: 0

**Reply by flaektrem** [reply]
Finally! I didn't get it to work until it changed name to Google Inc instead of Global Unichip. I don't know what i did.


````
root@Ubuntu:~/coral/tflite/python/examples/classification# python3 classify_image.py \
obilenet_v2_1.> --model models/mobilenet_v2_1.0_224_inat_bird_quant_edgetpu.tflite \
bels models/inat_bird_labels.txt \
--input images/parrot.jpg> --labels models/inat_bird_labels.txt \
> --input images/parrot.jpg
----INFERENCE TIME----
Note: The first inference on Edge TPU is slow because it includes loading the model into Edge TPU memory.
201.2ms
17.5ms
15.1ms
15.8ms
18.0ms
-------RESULTS--------
Ara macao (Scarlet Macaw): 0.77734
````

Now i've wasted several hours on this shit so it would be fun to know how to solve this quickly next time it happens.

**Reply by andynbaker** [reply]
Yeah I that sounds like what I've seen. It seems to have stabilized on my end and only changed after rebooting the host.

**Reply by towerhand** [reply]
The Edge TPU runtime and the  PyCoral library should be installed on the Proxmox host or the frigate LXC? Looks like inside the frigate lxc container reading the instructions on this discussion but https://github.com/blakeblackshear/frigate/discussions/1364 states to install them on the Proxmox host. Can someone please clarify? I'm able to run the model on the host if I install everything there but I can't get it to run the model inside the container.

**Reply by TheSleepySlee** [reply]
I did not install anything in the LXC, nor on the host. Just LXC
configuration. All the TPU drivers are in the docker image.

**Reply by towerhand** [reply]
So we don't need to do step 3 anymore? Only passthrough the USB device (step 4 - 5) in the /etc/pve/lxc/<container_id>.conf, even though it comes up as a Global Unichip (lsusb) and not as  Google Inc.?

**Reply by blakeblackshear** [reply]
Yep. The USB version doesn't require anything to be installed on the host.

**Reply by towerhand** [reply]
Thank you for the clarification.

**Comment by flaektrem** [comment]
Hi,
Followed your guide on LXC Ubuntu 20.04 but getting these errors while running the parrot test:
```
root@Ubuntu:~/coral/tflite/python/examples/classification# python3 classify_image.py --model models/mobilenet_v2_1.0_224_inat_bird_quant_edgetpu.tflite --labels models/inat_bird_labels.txt --input images/parrot.jpg
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/tflite_runtime/interpreter.py", line 152, in load_delegate
    delegate = Delegate(library, options)
  File "/usr/local/lib/python3.8/dist-packages/tflite_runtime/interpreter.py", line 111, in __init__
    raise ValueError(capture.message)
ValueError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "classify_image.py", line 122, in <module>
    main()
  File "classify_image.py", line 99, in main
    interpreter = make_interpreter(args.model)
  File "classify_image.py", line 72, in make_interpreter
    tflite.load_delegate(EDGETPU_SHARED_LIB,
  File "/usr/local/lib/python3.8/dist-packages/tflite_runtime/interpreter.py", line 154, in load_delegate
    raise ValueError('Failed to load delegate from {}\n{}'.format(
ValueError: Failed to load delegate from libedgetpu.so.1
```

Everyting is fine according to this:
https://coral.ai/docs/notes/build-coral/#failed-to-load-delegate-from-libedgetpuso1

Any clue? I have also tried to switch USB cable but no success.

Thanks

**Reply by thundercat49** [reply]
i had about given up on this ,desipte getting the LXC set up ( great instructions @andynbaker  - thankyou ) as i could never see "google" in my usb devices, i even tested the power out of the USB 3 port with a USB port tester despite trying everything - I switched the USB cable, and then left it alone for an hour , came back an hour later and its now there showing google, so dont know whether it was the cable swap or something else, but it didnt update immediately to "google" - must have spent 4 hours just figuring the USB side - trying to get it to pop up with google

still it is there now, and currently working with the HA add-in - so now i will continue to set up frigate in an LXC , hopefully tomorrow

**Reply by andynbaker** [reply]
So strange. Im starting to play with exporting promox data to influxdb that way I can easily keep an eye on system health. I'm going to set up some alerts if CPU load spikes.

**Reply by thundercat49** [reply]
im wondering if the coral USB is auto suspending either through a kernel setting on the proxmox OS , or  going to sleep via a setting within the underlying bios on my intel NUC ( im not a developer) , but that may well explain why it works for some people and not for others, ive changed the USB cable again to a much beefier cable, hasnt come back as yet, still shows unichip, but it has swapped usb path again, after a unplug/ re-plug

**Reply by andynbaker** [reply]
Mine has been super stable as long as the container is up. I'm using the stock USB cable on a USB 3.0 port.

And I'm on an old SFF hp desktop. 4th gen i5. Nothing special. I did enable virtualization in the bios but not sure if that has an effect on the lxc.

**Reply by TheSleepySlee** [reply]
Same. Super stable since I run docker on Lxc and not VM on proxmox.

**Comment by wpeterw** [comment]
Where do you store the clips/recordings ? Inside the container ? Or is it possible to mount an NFS share ?

**Reply by TheSleepySlee** [reply]
I mount nfs share on my NAS.

**Reply by wpeterw** [reply]
Do you mount this share from inside the container ? If so, how ? When I try to mount an nfs share I get this error:
`mount: only root can do that (effective UID is 100000)`

**Reply by TheSleepySlee** [reply]
I use root. And mount from container, not inside docker image. I simply use volume for docker image.

**Reply by flaektrem** [reply]
> 
> 
> I use root. And mount from container, not inside docker image. I simply use volume for docker image.

I'm looking for a fully working NFS solution for my recordings. Can you paste your docker-compose file so I understand how you solved it? 
Thank you

**Reply by TheSleepySlee** [reply]
Sorry for late response.
NFS is not really related to docker-compose.
What I had done is:

1. Create NFS mount on docker machine (for example on /opt/frigate/media)
2. Add this /opt/frigate/media as a volume in docker-compose
3. Use this volume in the frigate config file to store clips

I stopped doing this because, I believe, of instabilities of frigate around clips retention at that time (meaning I had a small local SSD which was getting full. The files were no longer written on the NFS since the mount disappeared and were written locally. I bought a bigger local disk which solved the problem.
